File: spring-security-jwt/src/main/java/com/tutorial/jwtsecurity/controller/MemberController.java
package com.tutorial.jwtsecurity.controller;

import com.tutorial.jwtsecurity.controller.dto.MemberResponseDto;
import com.tutorial.jwtsecurity.service.MemberService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

@RestController
@RequiredArgsConstructor
@RequestMapping("/member")
public class MemberController {
  private final MemberService memberService;

  @GetMapping("/me")
  public ResponseEntity<MemberResponseDto> getMyMemberInfo() {
    return ResponseEntity.ok(memberService.getMyInfo());
  }

  @GetMapping("/{email}")
  @PreAuthorize("hasPermission(#email,'ROLE_ADMIN')")
  public ResponseEntity<MemberResponseDto> getMemberInfo(@PathVariable String email) {
    return ResponseEntity.ok(memberService.getMemberInfo(email));
  }
}

File: spring-webflux-kotlin/src/main/kotlin/com/bcp0109/springwebfluxkotlin/mono/MemberController.kt
package com.bcp0109.springwebfluxkotlin.mono

import com.bcp0109.springwebfluxkotlin.coroutine.corouter.CoRouterController
import com.bcp0109.springwebfluxkotlin.domain.Member
import org.omg.CORBA.ServerRequest
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*
import org.springframework.web.client.RestTemplate
import org.springframework.web.reactive.function.client.WebClient
import reactor.core.publisher.Mono

@Controller
@RequestMapping("/v1")
class MemberController(
    private val memberService: MemberService
) {
    val log: Logger = LoggerFactory.getLogger(MemberController::class.java)

    @GetMapping("/home")
    fun home(): Mono<ResponseEntity<String>> {
        log.info("home request")
        return Mono.just(
            ResponseEntity.ok().body("Home")
        )
    }

    @GetMapping("/google")
    fun google(): Mono<String> {
        val template = RestTemplate()
        val response = template.getForObject("https://www.google.com", String::class.java)
        return Mono.just(response!!)
    }

    @GetMapping("/api/mine")
    fun google2(): Mono<ResponseEntity<String>> {
        log.info("mine request")
        val webClient = WebClient.builder().build()

        return webClient.get()
            .uri("http://localhost:8181/mine")
            .retrieve()
            .bodyToMono(String::class.java)
            .map {
                ResponseEntity.ok().body(it)
            }
    }

    @GetMapping("/signup")
    fun signup(
        @RequestParam name: String,
        @RequestParam age: Int
    ): Mono<ResponseEntity<Member>> {
        return memberService.signup(name, age)
            .flatMap {
                Mono.just(ResponseEntity.ok(it))
            }
    }

    @GetMapping("/members/{memberId}")
    fun findMember(
        @PathVariable("memberId") memberId: Long
    ): Mono<ResponseEntity<Member>> {
        return memberService.findMemberById(memberId)
                .flatMap {
                    Mono.just(ResponseEntity.ok(it))
                }
    }
}
